src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceWriter.cs

```diff
@@ -1644,6 +1679,31 @@ private void WriteRedFactory(Node nd)
 
             WriteLine(").CreateRed();");
             WriteLine("    }");
+
+            this.WriteLine();
+
+            // Add partial definition for a custom validator
+            Write("    static partial void Validate{0}Parts(", StripPost(nd.Name, "Syntax"));
+            WriteRedFactoryParameters(nd);
+            WriteLine(");");
+        }
+
+        private void WriteRedFactoryParameters(Node nd)
+        {
+            if (nd.Kinds.Count > 1)
+            {
+                Write("SyntaxKind kind, ");
+            }
+
+            for (int i = 0, n = nd.Fields.Count; i < n; i++)
+            {
+                var field = nd.Fields[i];
+                if (i > 0)
+                    Write(", ");
+                var type = this.GetRedPropertyType(field);
+
+                Write("{0} {1}", type, CamelCase(field.Name));
+            }
```

Comment:
This is from Aleksey's code in outvar branch.
I'm actually going to remove it from here, as I realized I don't need the `Internal="true"` anymore (your question above).